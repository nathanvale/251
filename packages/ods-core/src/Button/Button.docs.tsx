import React from "react";
import { ComponentDocs } from "@origin-digital/ods-types";
import { IconCheck } from "@origin-digital/ods-icons";
import { Stack } from "../Stack/Stack";
import { Columns } from "../Columns/Columns";
import { Column } from "../Column/Column";
import { Box } from "../Box/Box";
import { Text } from "../Text/Text";
import { Button, ButtonProps } from "./Button";

export const docs: ComponentDocs<ButtonProps> = {
  category: "Interaction",
  componentName: "Button",
  description: [
    "This is the atomic Button component fully styled according to Origin design system.",
  ].join(" "),
  propDescriptions: {
    buttonRef:
      "A React Ref object that allows accessing the underlying element",
    children: "The text to show on the button",
    color: [
      "The button could either of 2 colours:",
      "(1) 'primary': picks up theme.palette.primary color;",
      "(2) 'secondary': picks up theme.palette.grey[500] color.",
    ].join(" "),
    component: "The type of React element underneath this component.",
    "data-id":
      "The Button identifier that could be used for tracking and e2e testing purposes",
    fullWidth:
      "This is a responsiveProp, When true the button will take up the whole width of its parent on the defined screens.",
    href:
      "If provided the underlying element will be automatically set to <a/> and the href value is passed to it.",
    icon: "The icon to be shown before the Button text.",
    id: `The unique id of the Button, useful for accessibility.`,
    inverse: [
      "Use this when the button is shown on a dark background.",
      "This will flip the color and backgroundColor/borderColor accordingly to make sure the button is properly visible",
    ].join(" "),
    noTextPadding: [
      `Only applicable to variant="text", cancels out the padding by a negative margin.`,
      `This allows the button to look in harmony with text above or below it.`,
    ].join(" "),
    size:
      "The size of the button. This property is NOT responsive. This size should be set to small when using the text variant.",
    spinner: [
      "When true, will disable the Button and show a spinner on top of it.",
      "If the button is already disabled (`disabled={true}`), it won't show the",
      "spinner even if `spinner={true}`",
    ].join(" "),
    variant:
      "The button could visually have 3 variations as defined by this value",
  },
  migrationGuide: false,
  examples: {
    default: {
      Code: () => <Button>Button</Button>,
    },
    additional: [
      {
        label: "Contained Button",
        Code: () => (
          <Stack space="medium">
            <Columns space="medium">
              <Column width="content">
                <Button>Primary</Button>
              </Column>
              <Column>
                <Button disabled>Primary</Button>
              </Column>
            </Columns>
            <Columns space="medium">
              <Column width="content">
                <Button color="secondary">Secondary</Button>
              </Column>
              <Column>
                <Button disabled color="secondary">
                  Secondary
                </Button>
              </Column>
            </Columns>
          </Stack>
        ),
        codeString: `<Box>
   <Button>Primary</Button>
   <Button disabled>Primary</Button>
   <Button color="secondary">Secondary</Button>
   <Button disabled color="secondary">Secondary</Button>
</Box>`,
      },
      {
        label: "Outlined Button",
        Code: () => (
          <Stack space="medium">
            <Columns space="medium">
              <Column width="content">
                <Button variant="outlined">Primary</Button>
              </Column>
              <Column>
                <Button disabled variant="outlined">
                  Primary
                </Button>
              </Column>
            </Columns>
            <Columns space="medium">
              <Column width="content">
                <Button variant="outlined" color="secondary">
                  Secondary
                </Button>
              </Column>
              <Column>
                <Button disabled variant="outlined" color="secondary">
                  Secondary
                </Button>
              </Column>
            </Columns>
          </Stack>
        ),
        codeString: `<Box>
  <Button variant="outlined">Primary</Button>
  <Button disabled variant="outlined">Primary</Button>
  <Button variant="outlined" color="secondary">Secondary</Button>
  <Button disabled variant="outlined" color="secondary">Secondary</Button>
</Box>`,
      },
      {
        label: "Text Button",
        Code: () => (
          <Stack space="medium">
            <Columns space="medium">
              <Column width="content">
                <Button variant="text" size="small">
                  Primary
                </Button>
              </Column>
              <Column>
                <Button disabled variant="text" size="small">
                  Primary
                </Button>
              </Column>
            </Columns>
            <Columns space="medium">
              <Column width="content">
                <Button variant="text" color="secondary" size="small">
                  Secondary
                </Button>
              </Column>
              <Column>
                <Button disabled variant="text" color="secondary" size="small">
                  Secondary
                </Button>
              </Column>
            </Columns>
          </Stack>
        ),
        codeString: `<Box>
  <Button variant="text">Primary</Button>
  <Button disabled variant="text">Primary</Button>
  <Button variant="text" color="secondary">Secondary </Button>
  <Button disabled variant="text" color="secondary">Secondary</Button>
</Box>`,
      },
      {
        label: "Text Button",
        description:
          "To left-align with text under or above it the padding is cancelled by negative margins",
        Code: () => (
          <Stack space="medium" alignX="left">
            <Button noTextPadding variant="text" size="small">
              Primary
            </Button>
            <Text>This text should left-align with the button above</Text>
          </Stack>
        ),
        codeString: `<Stack space="medium" alignX="left">
  <Button noTextPadding variant="text" size="small">
    Primary
  </Button>
  <Text>This text should left-align with the button above</Text>
</Stack>`,
      },
      {
        label: "Link Button",
        description: "A button with href will render an <a /> tag.",
        Code: () => (
          <Button
            variant="text"
            href="https://www.originenergy.com.au"
            target="_blank"
            size="small"
          >
            Origin website
          </Button>
        ),
      },
      {
        label: "FullWidth Button",
        description: "A button that takes up the full width of its container.",
        Code: () => (
          <Stack alignX="left" space="small">
            <Button fullWidth>Full width</Button>
            <Button fullWidth={[true, false]}>Full width on xs, sm, md</Button>
            <Button
              fullWidth={{
                xs: true,
                sm: true,
                md: false,
                lg: false,
                xl: false,
              }}
            >
              Full width on xs, sm
            </Button>
          </Stack>
        ),
      },
      {
        label: "Button with Icon",
        description: "Button can render an icon before the text",
        Code: () => (
          <Box>
            <Button icon={<IconCheck />}>Medium</Button>
            <Box style={{ display: "inline-block", width: "20px" }} />
            <Button size="small" icon={<IconCheck />}>
              Small
            </Button>
          </Box>
        ),
        playroom: false,
        codeString: `<Box>
  <Button icon={<IconCheck />}>Medium</Button>
  <Box style={{ display: "inline-block", width: "20px" }} />
  <Button size="small" icon={<IconCheck />}>
    Small
  </Button>
</Box>`,
      },
      {
        label: "Button sizes",
        description:
          "Buttons can have 2 different sizes, 'medium' being the default.",
        Code: () => (
          <>
            <Box>
              <Button size="small">Small</Button>
              <Box style={{ display: "inline-block", width: "20px" }} />
              <Button size="medium">Medium</Button>
            </Box>
          </>
        ),
        codeString: `<Box>
  <Button size="small">Small</Button>
  <Button size="medium">Medium</Button>
</Box>`,
      },
      {
        label: "Inverse Button",
        description:
          "Button can swap out its color and background-color if is rendered on a dark background",
        Code: () => (
          <Box backgroundColor="grey600" height="full" padding="medium">
            <Columns space="small">
              <Column>
                <Stack space="small">
                  <Button inverse>Contained</Button>
                  <Button inverse variant="outlined">
                    Outlined
                  </Button>
                  <Button inverse variant="text">
                    Text
                  </Button>
                  <Button inverse disabled>
                    Disabled
                  </Button>
                  <Button inverse disabled variant="text">
                    Disabled
                  </Button>
                </Stack>
              </Column>
              <Column>
                {" "}
                <Stack space="small">
                  <Button inverse color="secondary">
                    Contained
                  </Button>

                  <Button inverse variant="outlined" color="secondary">
                    Outlined
                  </Button>

                  <Button inverse variant="text" color="secondary">
                    Text
                  </Button>

                  <Button inverse disabled variant="outlined" color="secondary">
                    Disabled
                  </Button>

                  <Button inverse disabled variant="text" color="secondary">
                    Disabled
                  </Button>
                </Stack>
              </Column>
            </Columns>
          </Box>
        ),
        codeString: `<Box>
  <Button inverse>Contained</Button>
  <Button inverse variant="outlined">Outlined</Button>
  <Button inverse variant="text">Text</Button>
  <Button inverse disabled>Disabled</Button>

  <Button inverse color="secondary">Contained</Button>
  <Button inverse variant="outlined" color="secondary">Outlined</Button>
  <Button inverse variant="text" color="secondary">Text</Button>
  <Button inverse disabled color="secondary">Disabled</Button>
</Box>`,
      },
      {
        label: "Button with action",
        description: "Button has onClick handler",
        // eslint-disable-next-line no-alert
        Code: () => <Button onClick={(e) => alert(e.target)}>Click !</Button>,
      },
      {
        label: "Button with spinner",
        description: "Turn spinner on",
        Code: () => <Button spinner>Submit</Button>,
      },
    ],
  },
  snippets: [
    {
      label: "Simple case - 2 Buttons",
      Code: () => (
        <Columns space="small">
          <Column width="content">
            <Button>Primary</Button>
          </Column>
          <Column>
            <Button color="secondary" variant="outlined">
              Secondary
            </Button>
          </Column>
        </Columns>
      ),
    },
    {
      label: "Full width",
      Code: () => <Button fullWidth>Submit form</Button>,
    },
    {
      label: "Full width array",
      Code: () => (
        <Button fullWidth={[true, false]}>Full width on small</Button>
      ),
    },
    {
      label: "Full width object",
      Code: () => (
        <Button
          fullWidth={{
            xs: true,
            sm: true,
            md: false,
            lg: false,
            xl: false,
          }}
        >
          Full width on small
        </Button>
      ),
    },
    {
      label: "Link Button",
      Code: () => (
        <Button
          variant="text"
          href="https://www.originenergy.com.au"
          target="_blank"
        >
          Origin website
        </Button>
      ),
    },
    {
      label: "Small",
      Code: () => <Button size="small">Small</Button>,
    },
    {
      label: "Text Button",
      Code: () => (
        <Columns space="small">
          <Column width="content">
            <Button variant="text">Primary</Button>
          </Column>
          <Column>
            <Button variant="text" color="secondary">
              Secondary
            </Button>
          </Column>
        </Columns>
      ),
    },
    {
      label: "Button with spinner",
      Code: () => <Button spinner>Submit</Button>,
    },
  ],
};
