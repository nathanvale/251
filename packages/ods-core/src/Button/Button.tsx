import React, {
  FocusEventHandler,
  KeyboardEventHandler,
  MouseEventHandler,
} from "react";
import {
  Button as MuiButton,
  ButtonProps as MuiButtonProps,
} from "@material-ui/core";
import {
  ButtonColor,
  ButtonElements,
  ButtonSize,
  ButtonVariant,
  InputType,
  MuiBasedComponentBaseProps,
  ResponsiveProp,
} from "@origin-digital/ods-types";
import { Box } from "../Box";
import { Spinner } from "../Spinner";
import { useTracking } from "../TrackingProvider";
import {
  useButtonStyles,
  useInverseStyles,
  useSpinnerStyles,
} from "./button-styles";

export interface ButtonActionProps {
  onFocus?: FocusEventHandler<HTMLButtonElement>;
  onBlur?: FocusEventHandler<HTMLButtonElement>;
  onClick?: MouseEventHandler<HTMLButtonElement>;
  onClickCapture?: MouseEventHandler<HTMLButtonElement>;
  onKeyDown?: KeyboardEventHandler<HTMLButtonElement>;
  onKeyPress?: KeyboardEventHandler<HTMLButtonElement>;
  onKeyUp?: KeyboardEventHandler<HTMLButtonElement>;
}

export interface ButtonProps
  extends Omit<MuiBasedComponentBaseProps, "classes">,
    ButtonActionProps {
  variant?: ButtonVariant;
  noTextPadding?: boolean;
  color?: ButtonColor; // secondary here is a grey colour.
  classes?: MuiButtonProps["classes"];
  size?: ButtonSize;
  fullWidth?: ResponsiveProp<boolean>;
  icon?: React.ReactNode; // it is always on the left.
  component?: ButtonElements;
  href?: string;
  target?: "_self" | "_blank" | string;
  inverse?: boolean;
  spinner?: boolean;
  type?: InputType;
  buttonRef?: React.Ref<unknown>;
  muiProps?: MuiButtonProps;
}

const defaultDataId = "button";

export const Button = (props: ButtonProps) => {
  const {
    children,
    color = "primary",
    component = "button",
    "data-id": dataId = defaultDataId,
    disabled,
    fullWidth,
    href,
    icon,
    inverse,
    muiProps,
    noTextPadding,
    spinner,
    size = "medium",
    type,
    variant = "contained",
    ...others
  } = props;

  const btnClasses = useButtonStyles({
    size,
    variant,
    hasIcon: !!icon,
    noTextPadding,
    fullWidth,
  });
  const inverseClasses = useInverseStyles({
    size,
    variant,
    hasIcon: !!icon,
    noTextPadding,
    fullWidth,
  });

  const spinnerClasses = useSpinnerStyles({ color, size, variant });

  const newComp = href ? "a" : component;

  const { onClickCapture, ref } = useTracking({
    children,
    "data-id": dataId,
    type: Button.displayName,
  });

  return (
    <MuiButton
      data-id={dataId}
      disabled={spinner || disabled}
      ref={ref}
      classes={inverse ? inverseClasses : btnClasses}
      color={color}
      component={newComp}
      href={href}
      type={type}
      size={size}
      startIcon={icon}
      variant={variant}
      onClickCapture={onClickCapture}
      {...others}
      {...muiProps}
    >
      {spinner && !disabled ? (
        <>
          <Box component="span" style={{ opacity: 0 }}>
            {children}
          </Box>
          <Spinner
            className={spinnerClasses.spinner}
            size={size === "medium" ? "small" : "xsmall"}
            data-id="button-spinner"
          />
        </>
      ) : (
        children
      )}
    </MuiButton>
  );
};

Button.displayName = "Button";

Button.defaultProps = {
  "data-id": defaultDataId,
  variant: "contained",
  size: "medium",
  color: "primary",
  component: "button",
};
