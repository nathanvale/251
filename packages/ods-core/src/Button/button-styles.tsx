import { makeStyles } from "@material-ui/core/styles";
import { Theme } from "@material-ui/core";
import {
  darken,
  fade,
  lighten,
} from "@material-ui/core/styles/colorManipulator";
import {
  ButtonColor,
  ButtonSize,
  ButtonVariant,
  ResponsiveProp,
} from "@origin-digital/ods-types";
import { Breakpoint } from "@material-ui/core/styles/createBreakpoints";
import { mapToStyledSystem } from "@origin-digital/ods-helpers";

const getHoverBG = (color: string) => {
  return darken(color, 0.16);
};
const getHoverBGInverse = (color: string) => {
  return lighten(color, 0.84);
};
const getHoverBGInverseFade = (color: string) => {
  return fade(color, 0.24);
};
const getDisabledBG = (color: string) => {
  return lighten(color, 0.68);
};
const getDisabled16BG = (color: string) => {
  return lighten(color, 0.84);
};

/**
 * This function receives the responsive with Prop and returns JSS responsive styles for the css "width" property.
 * @param theme
 * @param fullWidthArray
 */
const getArrayFullWidthStyles = (
  theme: Theme,
  fullWidthArray: (boolean | null)[]
) => {
  return theme.breakpoints.keys.reduce(
    (curStyles, bp: Breakpoint, index: number) => {
      curStyles[theme.breakpoints.up(bp)] = {
        width: fullWidthArray[index] ? "100%" : "auto",
      };
      return curStyles;
    },
    {} as any
  );
};

/**
 * fullWidth Prop is responsive. This function returns correct responsive `width` in the styles based on
 * the value of fullWidth (whether it is boolean, boolean[] or Object format).
 * The style is then used in the root style rule to apply the width on the button.
 * @param theme
 * @param fullWidth
 */
const getFullWidthStyles = (
  theme: Theme,
  fullWidth?: ResponsiveProp<boolean>
) => {
  if (!fullWidth) {
    return {
      width: "auto",
    };
  }
  const normFullWidth = mapToStyledSystem(fullWidth);
  if (typeof normFullWidth === "boolean") {
    return {
      width: normFullWidth ? "100%" : "auto",
    };
  } else if (normFullWidth instanceof Array) {
    return getArrayFullWidthStyles(theme, normFullWidth);
  } else {
    return {
      width: "auto",
    };
  }
};

/**
 * These root styles are common between normal and inverse buttons. It includes below logics:
 * - taking care of responsive fullWidth prop
 * - Padding logic:
 *   - based on the size (small / large)
 *   - icon is rendered (padding left is reduced)
 *   - variant outline has 1px smaller paddings to cater for borders
 * - Margin top / left: When the variant is text, we would like the text of the button to align with the rest of the
 * page. Hence we need to cancel out the left padding. Similarly the top padding. This is done by adding negative
 * margins when the variant="text" only. It still appreciates Layout components (e.g. Column and Stack) in the sense
 * that they can override these margins if they need to.
 * @param theme
 */
const getCommonStyleRules = (theme: Theme) => ({
  root: ({
    size,
    variant,
    hasIcon,
    noTextPadding,
    fullWidth,
  }: {
    size: ButtonSize;
    variant: ButtonVariant;
    hasIcon: boolean;
    noTextPadding?: boolean;
    fullWidth?: ResponsiveProp<boolean>;
  }) => {
    const fullWidthStyles = getFullWidthStyles(theme, fullWidth);

    //Handle padding for different button sizes and whether there is an icon or not.
    let [pt, pr, pb, pl] = [12, 24, 12, 24];
    if (size === "small" && hasIcon) {
      [pt, pr, pb, pl] = [12, 12, 12, 8];
    } else if (size === "small" && !hasIcon) {
      [pt, pr, pb, pl] = [12, 16, 12, 16];
    } else if (size === "medium" && hasIcon) {
      [pt, pr, pb, pl] = [12, 20, 12, 16];
    }

    // Outline has a border, so we need to subtract 1 pixel off each padding to have overall width of the button the same
    // as contained and text variants.
    if (variant === "outlined") {
      [pt, pr, pb, pl] = [pt - 1, pr - 1, pb - 1, pl - 1];
    }

    // Make sure we provide a specific CSS selector for the button label so that global resets do not override it.
    // Refer to https://origindd.atlassian.net/browse/ODS-221 for details.
    const buttonTypography = {
      "& > span.MuiButton-label": {
        fontWeight: theme.typography.button.fontWeight,
      },
    };

    if (noTextPadding && variant === "text") {
      return {
        padding: `${pt}px ${pr}px ${pb}px ${pl}px`,
        marginLeft: `-${pl}px`,
        marginRight: `-${pr}px`,
        marginTop: `-${pb}px`,
        marginBottom: `-${pb}px`,
        ...fullWidthStyles,
        ...buttonTypography,
      };
    }

    return {
      padding: `${pt}px ${pr}px ${pb}px ${pl}px`,
      ...fullWidthStyles,
      ...buttonTypography,
    };
  },
  sizeSmall: {
    fontSize: "14px",
    lineHeight: "16px",
  },
  iconSizeSmall: {
    "& > *:first-child": {
      fontSize: "16px",
      width: 16,
      height: 16,
    },
  },
});

export const useButtonStyles = makeStyles(
  (theme: Theme) => {
    return {
      ...getCommonStyleRules(theme),
      contained: {
        color: theme.palette.common.white,
        "&$disabled": {
          color: theme.palette.common.white,
        },
      },
      containedPrimary: {
        backgroundColor: theme.palette.primary.main,
        "&:hover": {
          backgroundColor: getHoverBG(theme.palette.primary.main),
        },
        "&$disabled": {
          backgroundColor: getDisabledBG(theme.palette.primary.main),
        },
      },
      containedSecondary: {
        backgroundColor: theme.palette.grey[500],
        "&:hover": {
          backgroundColor: getHoverBG(theme.palette.grey[500]),
        },
        "&$disabled": {
          backgroundColor: getDisabledBG(theme.palette.grey[500]),
        },
      },

      outlined: {
        backgroundColor: "transparent",
        "&$disabled": {
          backgroundColor: "transparent",
        },
      },
      outlinedPrimary: {
        "&$disabled": {
          color: getDisabledBG(theme.palette.primary.main),
          border: `1px solid ${getDisabled16BG(theme.palette.primary.main)}`,
        },
      },
      outlinedSecondary: {
        color: theme.palette.grey[500],
        border: `1px solid ${fade(theme.palette.grey[500], 0.5)}`,
        "&:hover": {
          border: `1px solid ${theme.palette.grey[500]}`,
          backgroundColor: fade(
            theme.palette.grey[500],
            theme.palette.action.hoverOpacity
          ),
        },
        "&$disabled": {
          color: getDisabledBG(theme.palette.grey[500]),
          border: `1px solid ${getDisabled16BG(theme.palette.grey[500])}`,
        },
      },

      textSecondary: {
        color: theme.palette.grey[500],
        "&:hover": {
          backgroundColor: fade(
            theme.palette.grey[500],
            theme.palette.action.hoverOpacity
          ),
        },
      },
      disabled: {},
    };
  },
  { classNamePrefix: "Button" }
);

export const useInverseStyles = makeStyles(
  (theme: Theme) => {
    return {
      ...getCommonStyleRules(theme),
      contained: {
        backgroundColor: theme.palette.common.white,
        "&$disabled": {
          color: theme.palette.grey[300],
          backgroundColor: darken(theme.palette.grey[500], 0.16),
        },
      },
      containedPrimary: {
        color: theme.palette.primary.main,
        "&:hover": {
          backgroundColor: getHoverBGInverse(theme.palette.primary.main),
        },
      },
      containedSecondary: {
        color: theme.palette.grey[500],
        "&:hover": {
          backgroundColor: getHoverBGInverse(theme.palette.grey[500]),
        },
      },

      outlined: {
        color: theme.palette.common.white,
        border: `1px solid ${fade(theme.palette.common.white, 0.5)}`,
        "&:hover": {
          border: `1px solid ${theme.palette.common.white}`,
          backgroundColor: getHoverBGInverseFade(theme.palette.grey[500]),
        },
        "&&$disabled": {
          color: theme.palette.grey[500],
          backgroundColor: "transparent",
          border: `1px solid ${theme.palette.grey[500]}`,
        },
      },

      text: {
        color: theme.palette.common.white,
        "&:hover": {
          backgroundColor: getHoverBGInverseFade(theme.palette.grey[500]),
        },
        "&$disabled": {
          color: theme.palette.grey[500],
        },
      },

      disabled: {},
    };
  },
  { classNamePrefix: "Button-inverse" }
);

export const useSpinnerStyles = makeStyles(
  (theme: Theme) => ({
    spinner: ({
      color = "primary",
      size = "medium",
      variant = "contained",
    }: {
      color: ButtonColor;
      size: ButtonSize;
      variant: ButtonVariant;
    }) => {
      let spinnerColor =
        color === "primary"
          ? theme.palette.primary.main
          : theme.palette.grey[500];

      if (variant === "contained") {
        spinnerColor = getHoverBG(spinnerColor);
      }

      return {
        color: spinnerColor,
        position: "absolute",
        top: "50%",
        left: "50%",
        marginTop: size === "small" ? -10 : -12,
        marginLeft: size === "small" ? -10 : -12,
      };
    },
  }),
  { classNamePrefix: "button" }
);
