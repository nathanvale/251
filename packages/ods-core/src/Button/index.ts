export { Button } from "./Button";
export type ButtonProps = import("./Button").ButtonProps;
export type ButtonActionProps = import("./Button").ButtonActionProps;
