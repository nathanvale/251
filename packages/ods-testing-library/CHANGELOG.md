# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [0.3.0](https://bitbucket.orgn.io/od/origin-ui/compare/@origin-digital/ods-testing-library@0.2.0...@origin-digital/ods-testing-library@0.3.0) (2020-02-12)


### Features

* **tests, columns:** alignY default changed to top ([c23c964](https://bitbucket.orgn.io/od/origin-ui/commits/c23c964))





# 0.2.0 (2020-01-31)


### Features

* set up jest for snapshots ([6e6fbfa](https://bitbucket.orgn.io/od/origin-ui/commits/6e6fbfa))
