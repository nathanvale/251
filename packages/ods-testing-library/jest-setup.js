import { configure } from "@testing-library/react";

require("jest-styled-components");

configure({ testIdAttribute: "data-id" });
