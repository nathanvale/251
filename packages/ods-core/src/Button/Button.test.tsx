import React from "react";
import { Phone } from "@material-ui/icons";
import { render, queryByAttribute } from "@origin-digital/ods-testing-library";
import { Button } from "./Button";

test("it renders a contained primary by default", () => {
  const { container } = render(
    <Button data-id="primary-button">Primary</Button>
  );

  expect(container).toMatchSnapshot();
});

test("it renders a contained secondary", () => {
  const { container } = render(<Button color="secondary">Primary</Button>);

  expect(container).toMatchSnapshot();
});

test("it renders a contained disabled", () => {
  const { container } = render(<Button disabled>Primary</Button>);

  expect(container).toMatchSnapshot();
});

test("it renders an outlined primary", () => {
  const { container } = render(<Button variant="outlined">Primary</Button>);

  expect(container).toMatchSnapshot();
});

test("it renders an outlined secondary", () => {
  const { container } = render(
    <Button variant="outlined" color="secondary">
      Primary
    </Button>
  );

  expect(container).toMatchSnapshot();
});

test("it renders an outlined disabled", () => {
  const { container } = render(
    <Button variant="outlined" disabled>
      Primary
    </Button>
  );

  expect(container).toMatchSnapshot();
});

test("it renders an text primary", () => {
  const { container } = render(<Button variant="text">Primary</Button>);

  expect(container).toMatchSnapshot();
});

test("it renders an text secondary", () => {
  const { container } = render(
    <Button variant="text" color="secondary">
      Primary
    </Button>
  );

  expect(container).toMatchSnapshot();
});

test("it renders an text disabled", () => {
  const { container } = render(
    <Button variant="text" disabled>
      Primary
    </Button>
  );

  expect(container).toMatchSnapshot();
});

test("it renders a small button", () => {
  const { container } = render(<Button size="small">Small</Button>);

  const buttonEl = container.querySelector("button");
  expect(buttonEl).toHaveClass("MuiButton-containedSizeSmall");
  expect(buttonEl).toHaveClass("MuiButton-sizeSmall");
});

test("it renders an icon", () => {
  const { container } = render(<Button icon={<Phone />}>Primary</Button>);

  expect(container).toMatchSnapshot();
});

test("it renders a primary inverse button", () => {
  const { container } = render(<Button inverse>Inverse</Button>);

  expect(container).toMatchSnapshot();
});

test("it renders a tag when href is provided", () => {
  const { container } = render(
    <Button data-id="my-button" href="https://www.originenergy.com.au">
      Origin
    </Button>
  );

  const buttonEl = queryByAttribute("data-id", container, "my-button");
  expect(buttonEl?.tagName).toEqual("A");
});

test("it calls onClick when clicked", () => {
  const onClick = jest.fn();
  const { container } = render(
    <Button data-id="my-button" onClick={onClick}>
      Origin
    </Button>
  );

  const buttonEl = queryByAttribute("data-id", container, "my-button");
  buttonEl?.click();
  expect(onClick).toHaveBeenCalledTimes(1);
});
