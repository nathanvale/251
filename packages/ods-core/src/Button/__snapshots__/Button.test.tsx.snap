// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders a contained disabled 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root Button-root-129 Button-root-140 MuiButton-contained Button-contained-132 MuiButton-containedPrimary Button-containedPrimary-133 Mui-disabled Button-disabled-139 Mui-disabled"
    disabled=""
    tabindex="-1"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Primary
    </span>
  </button>
</div>
`;

exports[`it renders a contained primary by default 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root Button-root-1 Button-root-12 MuiButton-contained Button-contained-4 MuiButton-containedPrimary Button-containedPrimary-5"
    data-id="primary-button"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Primary
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</div>
`;

exports[`it renders a contained secondary 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root Button-root-65 Button-root-76 MuiButton-contained Button-contained-68 MuiButton-containedSecondary Button-containedSecondary-70"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Primary
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</div>
`;

exports[`it renders a primary inverse button 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root Button-inverse-root-696 Button-inverse-root-705 MuiButton-contained Button-inverse-contained-699 MuiButton-containedPrimary Button-inverse-containedPrimary-700"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Inverse
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</div>
`;

exports[`it renders an icon 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root Button-root-611 Button-root-622 MuiButton-contained Button-contained-614 MuiButton-containedPrimary Button-containedPrimary-615"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      <span
        class="MuiButton-startIcon MuiButton-iconSizeMedium"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M6.62 10.79c1.44 2.83 3.76 5.14 6.59 6.59l2.2-2.2c.27-.27.67-.36 1.02-.24 1.12.37 2.33.57 3.57.57.55 0 1 .45 1 1V20c0 .55-.45 1-1 1-9.39 0-17-7.61-17-17 0-.55.45-1 1-1h3.5c.55 0 1 .45 1 1 0 1.25.2 2.45.57 3.57.11.35.03.74-.25 1.02l-2.2 2.2z"
          />
        </svg>
      </span>
      Primary
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</div>
`;

exports[`it renders an outlined disabled 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root Button-root-311 Button-root-322 MuiButton-outlined Button-outlined-317 MuiButton-outlinedPrimary Button-outlinedPrimary-318 Mui-disabled Button-disabled-321 Mui-disabled"
    disabled=""
    tabindex="-1"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Primary
    </span>
  </button>
</div>
`;

exports[`it renders an outlined primary 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root Button-root-183 Button-root-194 MuiButton-outlined Button-outlined-189 MuiButton-outlinedPrimary Button-outlinedPrimary-190"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Primary
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</div>
`;

exports[`it renders an outlined secondary 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root Button-root-247 Button-root-258 MuiButton-outlined Button-outlined-253 MuiButton-outlinedSecondary Button-outlinedSecondary-255"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Primary
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</div>
`;

exports[`it renders an text disabled 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root Button-root-493 Button-root-504 MuiButton-text MuiButton-textPrimary Mui-disabled Button-disabled-503 Mui-disabled"
    disabled=""
    tabindex="-1"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Primary
    </span>
  </button>
</div>
`;

exports[`it renders an text primary 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root Button-root-365 Button-root-376 MuiButton-text MuiButton-textPrimary"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Primary
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</div>
`;

exports[`it renders an text secondary 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root Button-root-429 Button-root-440 MuiButton-text MuiButton-textSecondary Button-textSecondary-438"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Primary
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</div>
`;
